### PR TITLE
Add a Commandlint that forces descriptions for commands and parameters

### PR DESCRIPTION
**Pull request by oakeyc**
Comes with a allowed-errors file that should decrease as the people add descriptions.
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] The PR has modified HISTORY.rst with an appropriate description of the change (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [x] Each command and parameter has a meaningful description.
- [x] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

File: scripts/automation/commandlint/run.py [hunk 1]
```diff
@@ -34,25 +37,35 @@ def dump_no_help(modules):
             print("EXCEPTION: " + str(mod))
 
     _update_command_definitions(cmd_table)
+    add_id_parameters(cmd_table)
+
+    with open(WHITE_DATA_FILE, 'r') as white:
+        white_data = json.load(white)
+
+    white_list_commands = white_data.get('commands', [])
```
Comment (troydai) [inline scripts/automation/commandlint/run.py]: Make it a set

File: scripts/automation/commandlint/run.py [hunk 2]
```diff
@@ -34,25 +37,35 @@ def dump_no_help(modules):
...
+    white_list_subgroups = white_data.get('subgroups', [])
```
Comment (troydai) [inline scripts/automation/commandlint/run.py]: Same

File: scripts/automation/commandlint/allowed-error.json [hunk 3]
```diff
@@ -0,0 +1,646 @@
...
+            "filters",
+            "expand"
+        ]
+    },
+    "subgroups": [
+        "appservice web config backup",
+        "resource link",
+        "appservice web deployment slot",
+        "webapp config backup",
+        "",
```
Comment (troydai) [inline scripts/automation/commandlint/allowed-error.json]: What is this for?

File: scripts/automation/commandlint/run.py [hunk 4]
```diff
@@ -85,6 +97,11 @@ def dump_no_help(modules):
...
     except ImportError:
         pass
-    exit_value = dump_no_help(installed_command_modules)
-    sys.exit(exit_value)
 
+    result, failed_commands = dump_no_help(installed_command_modules)
+
+    if any(len(failed_commands[key]) > 0 for key in failed_commands) or result != 0:
+        print('==== FAILED COMMANDS ====')
+        print(json.dumps(failed_commands, sort_keys=True, indent=4))
```
Comment (troydai) [inline scripts/automation/commandlint/run.py]: make indent 2 for a concise display.

File: scripts/build.sh [hunk 5]
```diff
@@ -23,6 +23,8 @@ if [ "$CODE_COVERAGE" == "True" ]; then
     pip install -qqq coverage codecov
     coverage run -m automation.tests.run
 
+    coverage run -m automation.commandlint.run
```
Comment (troydai) [inline scripts/build.sh]: Why do you need to run it under code coverage?

File: scripts/automation/commandlint/run.py (hunk 1 above)
Comment (oakeyc) [inline scripts/automation/commandlint/run.py]: Make the json a set?

File: scripts/build.sh (hunk 5 above)
Comment (oakeyc) [inline scripts/build.sh]: I don't, I changed it to just be a python command

File: scripts/automation/commandlint/run.py (hunk 4 above)
Comment (oakeyc) [inline scripts/automation/commandlint/run.py]: ack

File: scripts/automation/commandlint/run.py (hunk 2 above)
Comment (oakeyc) [inline scripts/automation/commandlint/run.py]: ack

File: scripts/automation/commandlint/allowed-error.json (hunk 3 above)
Comment (oakeyc) [inline scripts/automation/commandlint/allowed-error.json]: removed

File: scripts/automation/commandlint/allowed-error.json [hunk 6]
```diff
@@ -0,0 +1,646 @@
...
+            "years",
+            "cert",
+            "role",
+            "password",
+            "create_cert"
+        ],
+        "network watcher troubleshooting start": [
+            "storage_account",
+            "resource_type",
+        "storage_path"
```
Comment (oakeyc) [inline scripts/automation/commandlint/allowed-error.json]: ack

File: scripts/automation/commandlint/run.py [hunk 7]
```diff
@@ -42,30 +43,30 @@ def dump_no_help(modules):
     with open(WHITE_DATA_FILE, 'r') as white:
         white_data = json.load(white)
 
-    white_list_commands = white_data.get('commands', [])
-    white_list_subgroups = white_data.get('subgroups', [])
+    white_list_commands = sets.Set(white_data.get('commands', []))
```
Comment (troydai) [inline scripts/automation/commandlint/run.py]: doesn't `set(white_data.get('commands', [])` work?

File: scripts/automation/commandlint/run.py [hunk 8]
```diff
@@ -102,6 +109,6 @@ def dump_no_help(modules):
 
     if any(len(failed_commands[key]) > 0 for key in failed_commands) or result != 0:
         print('==== FAILED COMMANDS ====')
-        print(json.dumps(failed_commands, sort_keys=True, indent=4))
+        print(json.dumps(failed_commands, default=set_default, indent=2))
```
Comment (troydai) [inline scripts/automation/commandlint/run.py]: why? what's the issue here?

File: scripts/automation/commandlint/run.py (hunk 7 above)
Comment (oakeyc) [inline scripts/automation/commandlint/run.py]: set() is only for python 3

File: scripts/automation/commandlint/run.py (hunk 8 above)
Comment (oakeyc) [inline scripts/automation/commandlint/run.py]: Sets are not json serializable

**Comment by troydai** [comment]
Please rebase and squash your changes.

**Comment by oakeyc** [comment]
@troydai is there somewhere I should update the history that I'm adding a commandlinter?

**Comment by tjprescott** [comment]
@oakeyc since the command linter is a development team aid, and not customer-facing, there is no need to make a history entry.

**Pull request merged**